### PR TITLE
Logarithmic gossip fan out

### LINKED ISSUE #608: Optimize gossip
At the moment we send network-wide broadcast messages (like relayed transactions, consensus messages and generated blocks) to all the peers that the node has connected to it. It seems to be highly redundant and not that efficient.

There are different approaches to this problem, like the one outlined in [this paper](http://ceur-ws.org/Vol-2344/short2.pdf) that uses `1.4*ln(S) + 9` formula (for the case of more than 30 nodes) or the one [mentioned here](https://github.com/clockworksoul/smudge) that calculates `int(2.5 * log(number of nodes) + 0.5)`, probably many more with slightly different coefficients, but they tend to agree in that the number of fan-out nodes should be logarithmic relative to the number of nodes in the network.

We could use that approach if we were to know the number of nodes, but the problem for us is that we don't. We only know the number of our direct peers, we probably can estimate the number of peers that they have (but not reliably, P2P only has `getaddr`, unlike RPC's `getpeers`), but it seems to be impossible to get some reliable number of nodes in the network. 

So it seems like at the moment the best option for us is to continue broadcasting things to all peers we have, but maybe this bug inspires someone to develop a better approach given the constraints that we have.

Related to #590.

### PR DESCRIPTION
**Pull request by roman-khimov**
Fixes #608, #2678. A new configuration option is added, `BroadcastFactor` that allows to configure the number of fan-out node between optimal (log-dependent on the size of the network, default) and all available peers.

### CODE REVIEW

File: pkg/network/server.go [hunk 1]
```diff
@@ -387,10 +387,28 @@ func (s *Server) ConnectedPeers() []string {
 // while itself dealing with peers management (handling connects/disconnects).
 func (s *Server) run() {
 	go s.runProto()
-	for {
-		if s.PeerCount() < s.MinPeers {
+	for loopCnt := 0; ; loopCnt++ {
+		var (
+			netSize = s.discovery.NetworkSize()
+			// "Optimal" number of peers.
+			optimalN = s.discovery.GetFanOut() * 2
```
Comment (AnnaShaleva) [inline pkg/network/server.go:394]: Why `optimalN` is Fout*2?

File: pkg/network/server.go [hunk 2]
```diff
@@ -387,10 +387,28 @@ func (s *Server) ConnectedPeers() []string {
...
+			// Starting up or going below the minimum -> quickly get many new peers.
 			s.discovery.RequestRemote(s.AttemptConnPeers)
+		} else if s.MinPeers > 0 && loopCnt%s.MinPeers == 0 && optimalN > peerN && optimalN < s.MaxPeers && optimalN < netSize {
+			// Having some number of peers, but probably can get some more, the network is big.
+			// It also allows to start picking up new peers proactively, before we suddenly have <s.MinPeers of them.
+			var connN = s.AttemptConnPeers
+			if connN > optimalN-peerN {
+				connN = optimalN - peerN
+			}
+			s.discovery.RequestRemote(connN)
```
Comment (AnnaShaleva) [inline pkg/network/server.go:409]: Have you check this strategy on the network with a large number of available peers (maybe on mainnet o testnet)? What happens in the long run?

File: pkg/network/discovery.go [hunk 3]
```diff
@@ -189,9 +192,15 @@ func (d *DefaultDiscovery) RegisterConnectedAddr(addr string) {
 	d.lock.Lock()
 	delete(d.unconnectedAddrs, addr)
 	d.connectedAddrs[addr] = true
+	d.updateNetSize()
 	d.lock.Unlock()
 }
 
+// updateNetSize updates network size estimation metric. Must be called under read lock.
+func (d *DefaultDiscovery) updateNetSize() {
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/network/discovery.go:244]: Irrelevant.

File: pkg/network/prometheus.go [hunk 4]
```diff
@@ -6,6 +6,14 @@ import (
 
 // Metric used in monitoring service.
 var (
+	estimatedNetworkSize = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Help:      "Estimated network size",
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/network/prometheus.go:11]: I'm not sure it's a good idea. The problem is that we can't be sure that every unconnected peer really exists. It's an untrusted data in general, we only know peer exists when we connect to it.

[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/network/prometheus.go:11]: You're right, but at the same time it's definitely not the estimated network size either.

File: pkg/network/discovery.go [hunk 5]
```diff
@@ -93,48 +85,80 @@ func (d *DefaultDiscovery) backfill(addrs ...string) {
...
-		d.badAddrs[addr] = true
-		delete(d.unconnectedAddrs, addr)
-		delete(d.goodAddrs, addr)
+	for _, seed := range d.seeds {
+		if addr == seed {
+			isSeed = true
+			break
+		}
+	}
+	if !isSeed {
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/network/discovery.go:157]: Yes, we can't forget them, they're the last resort if all connections break.

File: pkg/network/server.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/network/server.go:394]: Multiplying by 2 can't be bad! It's all heuristics, there is no way around it. But it seems to be legit to me, we have ~100-200 nodes on the mainnet, so the fan out magic number is around 12-13 and we want ~20-30 connections to have good connectivity.

File: pkg/network/discovery.go [hunk 6]
```diff
@@ -93,48 +85,80 @@ func (d *DefaultDiscovery) backfill(addrs ...string) {
...
 // is already full, it just drops it.
 func (d *DefaultDiscovery) pushToPoolOrDrop(addr string) {
-	select {
-	case d.pool <- addr:
-		updatePoolCountMetric(d.PoolCount())
-		// ok, queued
-	default:
-		// whatever
+	if len(d.unconnectedAddrs) < maxPoolSize {
+		d.unconnectedAddrs[addr] = connRetries
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/network/discovery.go:110]: It's only called from `backfill()` now and it updates metrics after it's done with all addresses.

File: pkg/network/server.go (hunk 2 above)
Comment (roman-khimov) [reply pkg/network/server.go:409]: It works.

File: pkg/network/prometheus.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/network/prometheus.go:11]: To me it is. We don't know and can't know any better. The only way to prove it is by connecting to all nodes and checking them, but then a day later this data will be obsolete (that's why I also want to deal with `minPoolCount` eventually).

File: pkg/network/server.go (hunk 2 above)
Comment (roman-khimov) [reply pkg/network/server.go:409]: BTW, I've also looked at https://ethernodes.org/history --- E-letter chain has ~5-10K of nodes, so the optimal fan out for it (as we've defined it) is around 21-23 and it has these parameters:
```
./geth --maxpeers VALUE: set maximum number of full node peers, default 25
./geth --lightpeers VALUE: set maximum number of light node peers, default 100
```

**Review by AnnaShaleva — APPROVED** [review]

**Pull request merged**